protostar/profiler/transaction_profiler.py

```diff
@@ -0,0 +1,226 @@
+            current_contract = runtime_sample.callstack[-1]
+            for smp in runtime_sample.profile.step_samples:
+                step_samples.append(
+                    Sample(
+                        value=smp.value,
+                        callstack=translate_callstack(
+                            current_contract, global_instructions, smp.callstack
+                        )
+                        + upper,
+                    )
+                )
+            translated_callstacks = translate_callstacks(
+                current_contract,
+                global_instructions,
+                runtime_sample.profile.contract_call_callstacks,
+            )
+            new_callstacks_from_upper_layer.append(
+                [c + upper for c in translated_callstacks]
+            )
+        callstacks_from_upper_layer = list(
+            itertools.chain(*new_callstacks_from_upper_layer)
+        )
+
+    for smp in samples[-1].profile.memhole_samples:
+        current_contract = samples[-1].callstack[-1]
+        memhole_samples.append(
+            Sample(
+                value=smp.value,
+                callstack=translate_callstack(
+                    current_contract, global_instructions, smp.callstack
+                ),
+            )
+        )
+
+    # We build a tree from callstacks
+    callstacks_from_upper_layer = translate_callstacks(
+        samples[-1].callstack[-1],
+        global_instructions,
+        samples[-1].profile.contract_call_callstacks,
+    )
+    for i in range(2, max_clst_len + 1):
+        samples_in_layer = [s for s in samples if len(s.callstack) == i]
+        new_callstacks_from_upper_layer: List[List[List[Instruction]]] = []
+        for upper, runtime_sample in zip(callstacks_from_upper_layer, samples_in_layer):
+            current_contract = runtime_sample.callstack[-1]
+            for smp in runtime_sample.profile.memhole_samples:
+                memhole_samples.append(
+                    Sample(
+                        value=smp.value,
+                        callstack=translate_callstack(
+                            current_contract, global_instructions, smp.callstack
+                        )
+                        + upper,
+                    )
+                )
+
+            translated_callstacks = translate_callstacks(
+                current_contract,
+                global_instructions,
+                runtime_sample.profile.contract_call_callstacks,
+            )
+            new_callstacks_from_upper_layer.append(
+                [c + upper for c in translated_callstacks]
+            )
+        callstacks_from_upper_layer = list(
+            itertools.chain(*new_callstacks_from_upper_layer)
+        )
+
+    return TransactionProfile(
+        functions=list(global_functions.values()),
+        instructions=list(global_instructions.values()),
+        step_samples=step_samples,
+        memhole_samples=memhole_samples,
+    )
```

Comment:
This function is way too long.

Reply:
Agreed, I slowly decomposing it (removed few block from the top in this PR) I will proceed further in the following PR